### PR TITLE
Fix dependencies on highres and 16-bit in plumbers and dreamweb

### PR DESCRIPTION
**Pull request by phcoder**
Dreamweb has undeclared dependency on highres. It's used only for titles. Downscale titles if necessary instead.
Plumbers uses highres only for the windows version. Again BMPs are simple to downscale and the windows version has no movies.
3DO version needs 16-bit but it's better to disable only 3do version if 16-bit is unavailable.
3DO and windows share almost no code and instead it's a mess of variables which are glorified platform markers. I've split windows and 3do versions of plumbers into separate classes

### CODE REVIEW

File: engines/plumbers/metaengine.cpp [hunk 1]
```diff
@@ -40,7 +41,17 @@ class PlumbersMetaEngine : public AdvancedMetaEngine {
 };
 
 Common::Error PlumbersMetaEngine::createInstance(OSystem *syst, Engine **engine, const ADGameDescription *desc) const {
-	*engine = new Plumbers::PlumbersGame(syst, desc);
+	if (desc->platform == Common::kPlatform3DO) {
+#ifdef USE_RGB_COLOR
+		*engine = new Plumbers::PlumbersGame3DO(syst, desc);
+#else
+		// Should never happen. So no need to translate
+		GUI::MessageDialog dialog("3DO Plumbers requires RGB support.");
```
[the remark this answers is not in the document]
Comment (phcoder) [reply engines/plumbers/metaengine.cpp]: Done

File: graphics/module.mk [hunk 2]
```diff
@@ -41,6 +41,7 @@ MODULE_OBJS := \
 	primitives.o \
 	renderer.o \
 	scalerplugin.o \
+	scaler/downscaler.o \
```
[the remark this answers is not in the document]
Comment (phcoder) [reply graphics/module.mk:44]: Done

File: graphics/scaler/downscaler.cpp [hunk 3]
```diff
@@ -62,11 +63,70 @@ void DownscaleAllByHalfTemplate(const uint8 *srcPtr, uint32 srcPitch, uint8 *dst
...
+				}
+
+				*dst++ = col;
+			}
+		}
+		return;
+	}
+
+	if (in->format == PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0)
+	    && out->format  == PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0)) {
```
[the remark this answers is not in the document]
Comment (phcoder) [reply graphics/scaler/downscaler.cpp]: Done

File: engines/plumbers/3do.cpp [hunk 4]
```diff
@@ -0,0 +1,703 @@
...
+
+	_mouseHiLite = getMouseHiLite();
+}
+
+void PlumbersGame3DO::startGraphics() {
+	_image = new Image::Cel3DODecoder();
+	_ctrlHelpImage = new Image::Cel3DODecoder();
+	_screenW = 320;
+	_screenH = 240;
+	Graphics::PixelFormat pf(2, 5, 5, 5, 1, 10,  5,  0, 15);
```
[the remark this answers is not in the document]
Comment (phcoder) [reply engines/plumbers/3do.cpp:483]: Yes, Also RGB332 could be used for 8-bit colors. But again, for another PR.

File: graphics/scaler/downscaler.cpp [hunk 5]
```diff
@@ -62,11 +63,82 @@ void DownscaleAllByHalfTemplate(const uint8 *srcPtr, uint32 srcPitch, uint8 *dst
...
 }
 
 #endif
+
+namespace {
+uint32 getBrightness(byte col, const byte *pal) {
+	return pal[3 * col] * pal[3 * col] + pal[3 * col + 1] * pal[3 * col + 1] + pal[3 * col + 2] * pal[3 * col + 2];
+}
+
+void DownscaleCLUT8ByHalf(const uint8 *srcPtr, uint32 srcPitch, uint8 *dstPtr, uint32 dstPitch, int maxX, int maxY, const byte *palette) {
```
Comment (sev-) [inline graphics/scaler/downscaler.cpp:80]: This must start with lower capital letter, e.g. `downscaleCLUT8ByHalf()`

File: engines/engine.cpp [hunk 6]
```diff
@@ -310,37 +310,49 @@ void initGraphicsModes(const Graphics::ModeList &modes) {
...
 
-void initGraphics(int width, int height, const Graphics::PixelFormat *format) {
-
-	g_system->beginGFXTransaction();
+int initGraphicsAny(const Graphics::ModeWithFormatList &modes) {
+	int candidate = -1;
+	OSystem::TransactionError gfxError = OSystem::kTransactionSizeChangeFailed;
+	int last_width = 0, last_height = 0;
 
+	for (candidate = 0; candidate < (int) modes.size(); candidate++) {
```
Comment (sev-) [inline engines/engine.cpp:318]: Please, do not insert space after type conversions, e.g. `(int)modes.size()`

File: engines/engine.cpp [hunk 7]
```diff
@@ -310,37 +310,49 @@ void initGraphicsModes(const Graphics::ModeList &modes) {
 	g_system->initSizeHint(modes);
 }
 
-void initGraphics(int width, int height, const Graphics::PixelFormat *format) {
-
-	g_system->beginGFXTransaction();
+int initGraphicsAny(const Graphics::ModeWithFormatList &modes) {
```
Comment (sev-) [inline engines/engine.cpp:313]: Please add some documentation for the method.

File: engines/plumbers/3do.cpp [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+#include "graphics/palette.h"
+#include "graphics/surface.h"
+
+#include "image/cel_3do.h"
+
+#include "video/3do_decoder.h"
+
+namespace Plumbers {
+
+	// TODO(3do):
```
Comment (sev-) [inline engines/plumbers/3do.cpp:51]: Odd indentation

File: engines/plumbers/3do.cpp [hunk 9]
```diff
@@ -0,0 +1,703 @@
...
+static const Common::JoystickButton cheatJoy[] = {
+	Common::JOYSTICK_BUTTON_DPAD_UP,
+	Common::JOYSTICK_BUTTON_DPAD_DOWN,
+	Common::JOYSTICK_BUTTON_DPAD_RIGHT,
+	Common::JOYSTICK_BUTTON_DPAD_LEFT,
+	Common::JOYSTICK_BUTTON_DPAD_DOWN,
+	Common::JOYSTICK_BUTTON_DPAD_RIGHT,
+	Common::JOYSTICK_BUTTON_X
+};
+}
```
Comment (sev-) [inline engines/plumbers/3do.cpp:214]: I'd add here something like `// end of anonymous namespace`

File: engines/plumbers/3do.cpp [hunk 10]
```diff
@@ -0,0 +1,703 @@
...
+	file.seek(0x20290);
+
+	for (int i = 0; i < 26; i++) {
+		char buf[16];
+		file.read(buf, 16);
+		uint32 x = file.readUint32BE();
+		uint32 y = file.readUint32BE();
+		uint32 w = file.readUint32BE();
+		uint32 h = file.readUint32BE();
+		Common::String shortName = Common::String(buf).substr(0, 5);
```
Comment (sev-) [inline engines/plumbers/3do.cpp:363]: Ugh, this seems to be dangerous. Please write `Common::String(buf, 5)`. If bug is not null-terminated, you will have problem here and `String(const char *str, uint32 len)` is more secure and does exactly what you wrote with substr()

File: engines/plumbers/configure.engine [hunk 11]
```diff
@@ -1,3 +1,3 @@
 # This file is included from the main "configure" script
 # add_engine [name] [desc] [build-by-default] [subengines] [base games] [deps]
-add_engine plumbers "Plumbers Don't Wear Ties" yes "" "" "highres 16bit"
+add_engine plumbers "Plumbers Don't Wear Ties" yes "" "" ""
```
Comment (sev-) [inline engines/plumbers/configure.engine:3]: You might write 
```
add_engine plumbers "Plumbers Don't Wear Ties" yes
```

File: engines/plumbers/metaengine.cpp [hunk 12]
```diff
@@ -21,7 +21,10 @@
 
 #include "base/plugins.h"
 
+#include "common/translation.h"
```
Comment (sev-) [inline engines/plumbers/metaengine.cpp:24]: Please add `POTFILES`

File: engines/plumbers/metaengine.cpp [hunk 13]
```diff
@@ -40,7 +43,17 @@ class PlumbersMetaEngine : public AdvancedMetaEngine {
 };
 
 Common::Error PlumbersMetaEngine::createInstance(OSystem *syst, Engine **engine, const ADGameDescription *desc) const {
-	*engine = new Plumbers::PlumbersGame(syst, desc);
+	if (desc->platform == Common::kPlatform3DO) {
+#ifdef USE_RGB_COLOR
+		*engine = new Plumbers::PlumbersGame3DO(syst, desc);
+#else
+		// Should never happen. So no need to translate
+		GUI::MessageDialog dialog(_("3DO Plumbers requires RGB support."));
```
Comment (sev-) [inline engines/plumbers/metaengine.cpp:51]: Please add `// I18N: Plumbers is the title of the game`

This is the comment that the translators will see on Weblate.

File: engines/plumbers/module.mk [hunk 14]
```diff
@@ -2,9 +2,14 @@ MODULE := engines/plumbers
 
 MODULE_OBJS = \
 	plumbers.o \
+	windows.o \
```
Comment (sev-) [inline engines/plumbers/module.mk:5]: Please sort this list alphabetically.

File: engines/plumbers/plumbers.cpp [hunk 15]
```diff
@@ -678,353 +388,8 @@ void PlumbersGame::initTables() {
...
-		}
-		scene._decisionChoices++;
-	}
-
-	_totScene = scPtr;
-}
-
 int PlumbersGame::getSceneNumb(const Common::String &sName) {
 	debugC(1, kDebugGeneral, "%s : %s", __FUNCTION__, sName.c_str());
-	if (sName == "miketest/sc04" && _cheatEnabled)
```
Comment (sev-) [inline engines/plumbers/plumbers.cpp:1025]: Why is this removed?

File: engines/plumbers/windows.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		_actions.push(ChangeScene);
+		return;
+	}
+
+	_showScoreFl = true;
+	_leftButtonDownFl = true;
+	_setDurationFl = false;
+	loadImage(_scenes[_curSceneIdx]._sceneName + "/" + _scenes[_curSceneIdx]._decisionBitmap);
+}
+}
```
Comment (sev-) [inline engines/plumbers/windows.cpp:144]: Please add `// End of namespace Plumbers`

File: graphics/pixelformat.h [hunk 17]
```diff
@@ -199,6 +199,14 @@ struct PixelFormat {
 		       aShift == fmt.aShift;
 	}
 
+	inline bool isCLUT8() const {
```
Comment (sev-) [inline graphics/pixelformat.h:202]: This is already present in master

File: graphics/scaler/downscaler.cpp [hunk 18]
```diff
@@ -62,11 +63,82 @@ void DownscaleAllByHalfTemplate(const uint8 *srcPtr, uint32 srcPitch, uint8 *dst
...
+	if ((in->format == PixelFormat(2, 5, 5, 5, 0, 10, 5, 0, 0)
+	     && out->format == PixelFormat(2, 5, 5, 5, 0, 10, 5, 0, 0))
+	    || (in->format == PixelFormat(2, 5, 5, 5, 0, 0, 5, 10, 0)
+		&& out->format == PixelFormat(2, 5, 5, 5, 0, 0, 5, 10, 0))) {
+		DownscaleAllByHalf((const byte *) in->getBasePtr(0, 0), in->pitch, (byte *) out->getBasePtr(0, 0), out->pitch,
+				   MIN<int>(in->w, out->w * 2), MIN<int>(in->h, out->h * 2), 555);
+		return;
+	}
+
+	error("Unsupported downscale format %s->%s", in->format.toString().c_str(), out->format.toString().c_str());
```
Comment (sev-) [inline graphics/scaler/downscaler.cpp:141]: Please add prefix `downscaleCLUT8ByHalf(): ` to the message

File: graphics/scaler/downscaler.cpp [hunk 19]
```diff
@@ -62,11 +63,82 @@ void DownscaleAllByHalfTemplate(const uint8 *srcPtr, uint32 srcPitch, uint8 *dst
...
+}
+
+}
```
Comment (sev-) [inline graphics/scaler/downscaler.cpp:144]: Please add `// end of namespace Graphics`

**Review by sev- — CHANGES_REQUESTED** [review]
Added a few notes

**Comment by sev-** [comment]
@phcoder do you have time for finishing it?

File: graphics/scaler/downscaler.cpp (hunk 5 above)
Comment (phcoder) [reply graphics/scaler/downscaler.cpp:80]: Done

File: engines/engine.cpp (hunk 6 above)
Comment (phcoder) [reply engines/engine.cpp:318]: Done

File: graphics/scaler/downscaler.cpp (hunk 18 above)
Comment (phcoder) [reply graphics/scaler/downscaler.cpp:141]: Done

File: graphics/scaler/downscaler.cpp (hunk 19 above)
Comment (phcoder) [reply graphics/scaler/downscaler.cpp:144]: Done

File: engines/plumbers/3do.cpp (hunk 8 above)
Comment (phcoder) [reply engines/plumbers/3do.cpp:51]: Fixed

File: engines/plumbers/3do.cpp (hunk 10 above)
Comment (phcoder) [reply engines/plumbers/3do.cpp:363]: Done

File: engines/plumbers/plumbers.cpp (hunk 15 above)
Comment (phcoder) [reply engines/plumbers/plumbers.cpp:1025]: This is moved to 3do override.

File: engines/plumbers/windows.cpp (hunk 16 above)
Comment (phcoder) [reply engines/plumbers/windows.cpp:144]: Done

File: engines/plumbers/3do.cpp (hunk 9 above)
Comment (phcoder) [reply engines/plumbers/3do.cpp:214]: Done

File: engines/plumbers/module.mk (hunk 14 above)
Comment (phcoder) [reply engines/plumbers/module.mk:5]: Done

File: engines/plumbers/metaengine.cpp (hunk 13 above)
Comment (phcoder) [reply engines/plumbers/metaengine.cpp:51]: Done

File: engines/plumbers/configure.engine (hunk 11 above)
Comment (phcoder) [reply engines/plumbers/configure.engine:3]: Done

File: engines/plumbers/metaengine.cpp (hunk 12 above)
Comment (phcoder) [reply engines/plumbers/metaengine.cpp:24]: Done

File: graphics/pixelformat.h (hunk 17 above)
Comment (phcoder) [reply graphics/pixelformat.h:202]: Done

File: engines/engine.cpp (hunk 7 above)
Comment (phcoder) [reply engines/engine.cpp:317]: Done

**Comment by sev-** [comment]
Thank you!
